src/contexts/AppController.tsx

```diff
@@ -0,0 +1,202 @@
+    const setUiState = useUiStore((state) => state.setUiState);
+    const setPostPopUp = useContextStore((state) => state.setPostPopUp);
+
+    // Refs
+    const isPrettifiedRef = useRef<boolean>(isPrettified ?? false);
+    const logEventNumRef = useRef(logEventNum);
+
+    // Synchronize `logEventNumRef` with `logEventNum`.
+    useEffect(() => {
+        if (null !== logEventNum) {
+            logEventNumRef.current = logEventNum;
+            setLogEventNum(logEventNum);
+        }
+    }, [
+        logEventNum,
+        setLogEventNum,
+    ]);
+
+    // Synchronize `isPrettifiedRef` with `isPrettified`.
+    useEffect(() => {
+        isPrettifiedRef.current = isPrettified ?? false;
+        setIsPrettified(isPrettifiedRef.current);
+    }, [
+        isPrettified,
+        setIsPrettified,
+    ]);
+
+    // On `logEventNum` update, clamp it then switch page if necessary or simply update the URL.
+    useEffect(() => {
+        if (0 === numEvents || URL_HASH_PARAMS_DEFAULT.logEventNum === logEventNum) {
+            return;
+        }
+
+        const logEventNumsOnPage: number [] =
+            Array.from(beginLineNumToLogEventNum.values());
+
+        const clampedLogEventNum = clamp(logEventNum, 1, numEvents);
+
+        if (updateUrlIfEventOnPage(clampedLogEventNum, logEventNumsOnPage)) {
+            // No need to request a new page since the log event is on the current page.
+            return;
+        }
+
+        const cursor: CursorType = {
+            code: CURSOR_CODE.EVENT_NUM,
+            args: {eventNum: logEventNum},
+        };
+
+        setUiState(UI_STATE.FAST_LOADING);
+
+        (async () => {
+            const pageData = await logFileManagerProxy.loadPage(cursor, isPrettifiedRef.current);
+            useViewStore.getState().updatePageData(pageData);
+        })().catch((e: unknown) => {
+            postPopUp({
+                level: LOG_LEVEL.ERROR,
+                message: String(e),
+                timeoutMillis: DO_NOT_TIMEOUT_VALUE,
+                title: "Action failed",
+            });
+        });
+    }, [
+        beginLineNumToLogEventNum,
+        logEventNum,
+        logFileManagerProxy,
+        numEvents,
+        setUiState,
+        postPopUp,
+    ]);
+
+    // On `filePath` update, load file.
+    useEffect(() => {
+        if (URL_SEARCH_PARAMS_DEFAULT.filePath === filePath) {
+            return;
+        }
+
+        let cursor: CursorType = {code: CURSOR_CODE.LAST_EVENT, args: null};
+        if (URL_HASH_PARAMS_DEFAULT.logEventNum !== logEventNumRef.current) {
+            cursor = {
+                code: CURSOR_CODE.EVENT_NUM,
+                args: {eventNum: logEventNumRef.current},
+            };
+        }
+        loadFile(filePath, cursor);
+    }, [
+        filePath,
+        loadFile,
+    ]);
+
+    useEffect(() => {
+        setPostPopUp(postPopUp);
+    }, [
+        postPopUp,
+        setPostPopUp,
+    ]);
```

Comment:
We should `setPostPopUp()` to update the reference in `contextStore` before loading a file, or any error occured during file loading would not show a UI pop-up. e.g., accessing this should show a 404 error pop-up but nothing was shown: https://y-scope.github.io/yscope-log-viewer/?filePath=does_not_exist

@coderabbitai create an issue to track this